notify/jira/jira.go

```diff
@@ -0,0 +1,339 @@
+	}}
+
+	issueDescriptionString, err := tmplTextFunc(n.conf.Summary)
+	if err != nil {
+		return issue{}, fmt.Errorf("template error: %w", err)
+	}
+
+	if strings.HasSuffix(n.conf.APIURL.Path, "/3") {
+		var issueDescription any
+		if err := json.Unmarshal([]byte(issueDescriptionString), &issueDescription); err != nil {
+			return issue{}, nil
+		}
+		requestBody.Fields.Description = issueDescription
+	} else {
+		requestBody.Fields.Description = issueDescriptionString
+	}
+
+	if n.conf.StaticLabels != nil {
+		requestBody.Fields.Labels = n.conf.StaticLabels
+	}
+
+	if n.conf.Components != nil {
+		requestBody.Fields.Components = make([]idNameValue, len(n.conf.Components))
+
+		for i, components := range n.conf.Components {
+			requestBody.Fields.Components[i] = idNameValue{Name: components}
+		}
+	}
+
+	priority, err := tmplTextFunc(n.conf.Priority)
+	if err != nil {
+		return issue{}, fmt.Errorf("template error: %w", err)
+	}
+
+	if priority != "" {
+		requestBody.Fields.Priority = &idNameValue{Name: priority}
+	}
+
+	sort.Strings(requestBody.Fields.Labels)
+
+	return requestBody, nil
+}
+
+func (n *Notifier) searchExistingIssue(tmplTextFunc templateFunc, key notify.Key, status model.AlertStatus) (*issue, bool, error) {
+	jql := strings.Builder{}
+
+	if n.conf.WontFixResolution != "" {
+		jql.WriteString(fmt.Sprintf(`resolution != "%s" and `, n.conf.WontFixResolution))
+	}
+
+	// if alert is firing, do not search for closed issues unless reopen transition is defined.
+	if n.conf.ReopenTransition == "" {
+		if status != model.AlertResolved {
+			jql.WriteString(`statusCategory != Done and `)
+		}
+	} else {
+		reopenDuration := int64(time.Duration(n.conf.ReopenDuration).Minutes())
+		if reopenDuration != 0 {
+			jql.WriteString(fmt.Sprintf(`(resolutiondate is EMPTY OR resolutiondate >= -%dm) and `, reopenDuration))
+		}
+	}
+
+	jql.WriteString(fmt.Sprintf(`project = "%s" and labels=%q order by status ASC,resolutiondate DESC`, n.conf.Project, key.Hash()))
+
+	requestBody := issueSearch{}
+	requestBody.Jql = jql.String()
+	requestBody.MaxResults = 2
+	requestBody.Fields = []string{"status"}
+	requestBody.Expand = []string{}
+
+	level.Debug(n.logger).Log("msg", "search for recent issues", "alert", key.String(), "jql", jql.String())
+
+	responseBody, shouldRetry, err := n.doAPIRequest(tmplTextFunc, http.MethodPost, "search", requestBody)
+	if err != nil {
+		return nil, shouldRetry, err
+	}
+
+	var issueSearchResult issueSearchResult
+	err = json.Unmarshal(responseBody, &issueSearchResult)
+	if err != nil {
+		return nil, false, err
+	}
+
+	switch issueSearchResult.Total {
+	case 0:
+		level.Debug(n.logger).Log("msg", "found no existing issue", "alert", key.String())
+		return nil, false, nil
+	default:
+		level.Warn(n.logger).Log("msg", "more than one issue matched, picking most recently resolved", "alert", key.String(), "picked", issueSearchResult.Issues[0].Key)
+
+		fallthrough
+	case 1:
+		return &issueSearchResult.Issues[0], false, nil
+	}
```

Comment:
I've found the `fallthrough` trick a bit hard to follow. I'd prefer a simpler version.

```suggestion
	if issueSearchResult.Total == 0 {
		level.Debug(n.logger).Log("msg", "found no existing issue", "alert", key.String())
		return nil, false, nil
	}
	
	if issueSearchResult.Total > 1 {
		level.Warn(n.logger).Log("msg", "more than one issue matched, selecting the most recently resolved", "alert", key.String(), "selected", issueSearchResult.Issues[0].Key)
	}

	&issueSearchResult.Issues[0], false, nil
```

Reply:
fixed suggestion with return at the final line

```suggestion
	if issueSearchResult.Total == 0 {
		level.Debug(n.logger).Log("msg", "found no existing issue", "alert", key.String())
		return nil, false, nil
	}
	
	if issueSearchResult.Total > 1 {
		level.Warn(n.logger).Log("msg", "more than one issue matched, selecting the most recently resolved", "alert", key.String(), "selected", issueSearchResult.Issues[0].Key)
	}
	
	return &issueSearchResult.Issues[0], false, nil
```